src/coreclr/vm/finalizerthread.cpp

```diff
@@ -28,6 +28,7 @@ Volatile<BOOL> g_TriggerHeapDump = FALSE;
 CLREvent * FinalizerThread::hEventFinalizer = NULL;
 CLREvent * FinalizerThread::hEventFinalizerDone = NULL;
 CLREvent * FinalizerThread::hEventFinalizerToShutDown = NULL;
+CLREvent * FinalizerThread::hEventManagedFinalizerActionDone = NULL;
```

Comment:
We should not be adding more global synchronization around finalizer thread. The problem that you are trying to fix should be fixable without it by separating the tracking of the work that you expect the finalizer to do from tracking of the current state of the system.

Reply:
@jkotas, I believe adding this synchronization to the finalizer thread is the right thing to do. 

In case this sequence of events happened. We are obligated to deliver the callback.

1. User creates a NoGCRegion
2. User registers a callback
3. User allocates memory, to the point that the NoGCRegion terminates, and
4. User starts a new NoGCRegion again without exception.

In this case, the user observed the fact that NoGCRegion is completely exhausted. Then it implies the allocation definitely crossed the threshold. So if the callback is not delivered, that is an observable bug.

With that fact, we cannot cancel the callback work. If the finalizer thread got stuck somehow, we have two choices, either wait until the finalizer thread completes the NoGCCallback work before allowing the next one to start (which is what the PR is currently doing), or maintain a queue of NoGCCallback work items, and deliver them in order later when the finalizer thread get unstuck.

The latter option is bad for a few reasons:

1. It is hard to use - going back to the motivating scenario for this API where the user wants to use the callback to terminate the game. A late callback might terminate the wrong game. Missing a callback before the next game starts also messes up with game logic. What if the callback says the current game's score should not be counted? 

2. It is not useful - a NoGCCallback for a NoGCRegion that is already terminated a few NoGCRegions before. What can the programmer do about that?

3. It is a lot of complexity to implement. We need to maintain that queue of work items somewhere (either in the GC or in the finalizer thread), even with that, the finalizer thread still needs to be synchronized with respect to the use of the queue. Things just get more complicated, no less.

It is so much easier for everyone if we synchronize the finalizer thread so that we don't need the queue of callbacks.

Reply:
> In this case, the user observed the fact that NoGCRegion is completely exhausted. Then it implies the allocation definitely crossed the threshold. So if the callback is not delivered, that is an observable bug.

I think waiting for the callback to be delivered in this case is worse perf bug. It is very likely that the app does not actually care about this callback to be delivered, and waiting for it would lead to unnecessary delay, depending on how much work is finalizer doing at the moment.

If you do not want to deal with queue of callbacks, I think it better to skip the delivery of the callback in this case rather than blocking for it to be delivered.

Reply:
I am more concerned with the correctness guarantee than perf. One possible code pattern that could happen is something like this.

```
StartNoGCRegion()
RegisterCallback(Compensate)
PerformWork()
if EndNoGCRegion failed <- I knew I allocated too much, so let's wait for the compensation to be completed.
    WaitUntilCompensationCompletes().
DoSomethingElse()
```

If we skip the callback - that app would deadlock, and that would not be good.

For performance, performance-critical work for an app that uses NoGCRegion usually doesn't start before the NoGCRegion is started. For example, the part of a game that needs responsiveness doesn't start until the NoGCRegion is started. We probably all experienced that - starting a game can take a while. The game itself does a lot of preloading to get started anyway. Waiting before the finalizers are drained before running the next game is probably a good thing to have.

Reply:
> If we skip the callback - that app would deadlock, and that would not be good.

We can document that the callback may not be delivered if the GC region that it was registered for ended. Writing async callbacks is hard.

> Waiting before the finalizers are drained before running the next game is probably a good thing to have.

It depends. We have an API for draining the finalizers and the app can choose to call the API if it wants to do that.

Reply:
> the callback may not be delivered if the GC region that it was registered for ended.

With that caveat, as the game developer, I would have a hard time using it right. The right thing to do is something like this.

```
StartNoGCRegion()
RegisterCallback(TermTheGame)
StartGame()
if EndNoGCRegion failed:
    If TermTheGame is not already called, 
        interlocked switch on a flag so TermTheGame will become a no-op. 
        If the Interlocked operation failed, meaning the finalizer thread wins the race, 
            wait until TermTheGame completes.
StartNextGame()
```

If I don't do that switch flag, the guarantee we have does not guarantee the callback won't get called while the next game is in progress, and that can potentially kill the wrong game.

If we don't wait until the `TermTheGame` finishes on the finalizer thread, `TermTheGame` might run concurrently with `StartNextGame`, which is not good either.

A potentially more useful guarantee can be this

> As long as we never call `TryStartNoGCRegion` after a NoGCRegion ends, the callback is always delivered. But if we call `TryStartNoGCRegion`, then it _might_ cancel the work.

Still, there is no guarantee that the callback and trying to start the next NoGCRegion happen at exactly the same time. So I still need to worry about the possibility that the `TermTheGame` runs concurrently with `StartNextGame`. So I still need to wait until the callback is delivered. However, with this new version, as long as I make sure I don't call `TryStartNoGCRegion`, the waiting code will never deadlock.

The advantage of this version is that from the implementation perspective, we can simply null the callback when we start a NoGCRegion, which means we can eliminate the finalizer thread synchronization. The _might_ above refers to the possibility that the finalizer thread race with the thread starting a new NoGCRegion and therefore still have a potential that the callback is executed while starting (and after starting) a new NoGCRegion - technically that implies the cancellation is not reliable. 

I agree, writing async callback is really hard ...

Including the people in the original API proposal to see if they have something to say about what guarantee they want to be useful.

@GSPP, @hez2010, @lucabol, @WenceyWang

Reply:
> With that caveat, as the game developer, I would have a hard time using it right. The right thing to do is something like this.

I would expect the callback to be an instance method on the game, each game instance to allocate a new callback, and the first thing in your callback would be a check that the game is disposed - if it is disposed, do nothing.

Reply:
That avoids the issue that the callback work being performed after `dispose`, you also need to wait until the callback completes if it is already in progress before `dispose`.

Reply:
I am sorry about this late response but after all, It's already hard enough for a new developer to use NoGCRegion.

```
As long as we never call TryStartNoGCRegion after a NoGCRegion ends, the callback is always delivered. But if we call TryStartNoGCRegion, then it might cancel the work.
```
This seems fine to me as if the user code is calling TryStartNoGCRegion it should already be aware of the end of NoGCRegion.

But I want to point out that the difficulty of the use of NoGCRegion is mainly caused by computing the amount of memory to allocate and the idea of NoGCRegion can be automatically canceled. I am still hoping there can be a ContinueAllocatingThen* mode.

Reply:
Thanks, @WenceyWang. With your feedback, I have removed the finalizer thread synchronization and implemented the code to cancel the callback in `TryStartNoGCRegion` instead.

With regard to your comment about computing the amount of memory to allocate, note that in .NET 7 or above (except x86 or OSX), the size of the NoGC region is no longer restricted by the size of an ephemeral segment.

Reply:
@jkotas, the finalization thread synchronization issue is resolved. Can you take another look at this PR?